### PR TITLE
Read python file out of libpath

### PR DESCRIPTION
**Pull request by alexsavio**

### CODE REVIEW

**Comment by tony** [comment]
@alexsavio thanks.
There appears to be test issues. Can you take a look

**Comment by alexsavio** [comment]
I didn't know you were supporting older versions of Python. I will check it later.

**Comment by alexsavio** [comment]
The only error now is with importlib in Python 3.2.
I don't have access to a Python 3.2.
The documentation is wrong: https://docs.python.org/3.2/library/importlib.html
Can anyone help me?

**Comment by tony** [comment]
I don’t have a setup for older python versions (at the moment at least)

However, you could use pyenv to test older python versions.

https://github.com/yyuu/pyenv-installer to install pyenv 
then you can install older python interpreters

```
pyenv install 3.2
pyenv global system 3.2
rm -rf ~/.pyenv/shims && pyenv rehash
```

then you should have python3.2 in your paths.  I think you could also do

``pyenv virtualenv 3.2 kaptan3.2``   (create the virtual env)
``pyenv activate kaptan3.2``      (activate virtualenv)

then you can ``pip install -e /path/to/kaptandir``, and anything you run inside that virtual environment will use the latest changes you have made in that directory.

-- 
Tony Narlock
tony@git-pull.com

On December 8, 2015 at 5:05:30 AM, Alexandre M. S. (notifications@github.com) wrote:

The only error now is with importlib in Python 3.2.
I don't have access to a Python 3.2.
The documentation is wrong: https://docs.python.org/3.2/library/importlib.html
Can anyone help me?

**Comment by alexsavio** [comment]
Thank you very much @tony, I learned something from your reply. :)
Please check the code, I think it works now.

File: kaptan/__init__.py [hunk 1]
```diff
@@ -52,20 +53,39 @@ def upsert(self, key, value):
...
     def import_config(self, value):
-        if not isinstance(value, dict) and os.path.isfile(value):
+        if isinstance(value, dict):  # load python dict
+            self.handler = self.HANDLER_MAP['dict']()
+            data = value
+        elif op.isfile(value) and not self._is_python_file(value):
             if not self.handler:
                 try:
-                    key = HANDLER_EXT.get(os.path.splitext(value)[1][1:], None)
+                    key = HANDLER_EXT.get(op.splitext(value)[1][1:], None)
```
Comment (tony) [inline kaptan/__init__.py]: Just keeping it as ``os.path`` is a bit more readable.  It's not immediately obvious what ``op`` is if someone is glancing through it (unless they remembered how ``os.path`` was imported as ``op`` above).  May as well make it simpler for them :smile:

File: kaptan/__init__.py [hunk 2]
```diff
@@ -52,20 +53,39 @@ def upsert(self, key, value):
         self.configuration_data.update({key: value})
         return self
 
+    def _is_python_file(self, value):
```
Comment (tony) [inline kaptan/__init__.py]: For consistency (and keeping with the pattern we want), we probably would want to keep "config type" (e.g. python, ini, json, yaml)-specific code in it's corresponding module.

For example:
We could create an 'extensions' property in each subclass of ``BaseHandler`` and check to see if that filetype is found. If so, pick that filetype's handler.

```python
class FileHandler(BaseHandler):

    def load(self, file_):
```
could have

```python
class FileHandler(BaseHandler):
    extensions = ['py']
    def load(self, file_):
```

Comment (tony) [inline kaptan/__init__.py]: Can we get some documentation as to what "value" is supposed to look like? i.e. is it a full file path as a string (``/path/to/file.ext``), or is it just the file (``file.ext``).

**Comment by tony** [comment]
Related issue #23

File: kaptan/__init__.py (hunk 2 above)
Comment (tony) [inline kaptan/__init__.py]: Regarding that refactor that idea, probably out of scope for this PR.  I'll make a separate issue for that to tackle at a later time.

**Comment by tony** [comment]
I'm going to give this a second look tomorrow.  Disregard that comment I made about refactoring file extensions - we already have ``HANDLER_EXT`` for that atm.

@alexsavio : To double check, https://github.com/emre/kaptan/issues/23#issue-120749487

In this issue, are you also doing the issue to load the handler purely by file extension? Or just importing python module part?

**Comment by alexsavio** [comment]
Hi @tony,

Kaptan is already loading the handler according to the file extension. The only problem that I found is that FileHandler needed the '.py' file to be in the PATH of the interpreter and also `file_` needed to be the name of the file without the `.py` extension. The later seemed weird for me when I tried Kaptan.

The name FileHandler is still a bit confusing for me, why not PyFileHandler? :)

Some docstrings all around the code would help a bit as well. I wouldn't mind doing that.

**Comment by tony** [comment]
> The name FileHandler is still a bit confusing for me, why not PyFileHandler? :)

It is quite ambiguous. PyFileHandler sounds nice!

If you have any other things you can find to improve code quality / maintainability, we could really use your help on that.

@alexsavio: I'm merging this in.

I'm going to be tidying up the comments / docs a bit after this as well.

**Pull request merged**